### PR TITLE
Don't panic on unexpected group version for mutated selector

### PR DESCRIPTION
**Pull request by janetkuo**
**What this PR does / why we need it**: 
`panic` on unknown group version will crash the whole apiserver (even when selector isn't mutated), have a single resource affect the whole cluster, and make debugging harder. ~Return an internal error (i.e. not user error) instead.~ Make "selector immutability" the default behavior instead. 

**Which issue(s) this PR fixes**:
ref #54575

**Special notes for your reviewer**: @kubernetes/sig-apps-bugs @liggitt 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
> panic on unknown group version will crash the whole apiserver (even when selector isn't mutated), have a single resource affect the whole cluster

that's true for all API server panics today, and is intentional (see https://github.com/kubernetes/kubernetes/pull/28800). I think it's a bad default for a multi-user request handler like the API server, but that's what has been decided.

File: staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go [hunk 1]
```diff
@@ -128,6 +128,16 @@ func ValidateImmutableField(newVal, oldVal interface{}, fldPath *field.Path) fie
 	return allErrs
 }
 
+// We don't know how to handle mutated selector for this unexpected group version.
+// Throw an internal error.
+func ValidateImmutableFieldForUnexpectedGroupVersion(newVal, oldVal interface{}, fldPath *field.Path, groupVersion string) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(ValidateImmutableField(newVal, oldVal, fldPath)) > 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go]: don't make this conditional on the field being mutated... it's a programming error to not make a decision about this and we don't want to ship a release with an unknown version in these cases

Comment (janetkuo) [reply staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go]: Is there a way to catch this programming error, when a new apiversion is added without making a decision in update strategy?

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go]: a unit test in the registry packages to check registered versions for the relevant objects would be my recommendation.

also, this isn't generic enough to warrant being in k8s.io/apimachinery

Comment (janetkuo) [reply staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go]: > a unit test in the registry packages to check registered versions for the relevant objects would be my recommendation.

Did you mean a unit test here, or something else?
https://github.com/kubernetes/kubernetes/blob/3a15fdbe7e1ceefa59d7cc41ac6d65998cef1b1a/pkg/registry/extensions/daemonset/strategy_test.go#L47-L73

If a new version is added without adding another test case here, the error still cannot be caught.

**Review by kow3ns — COMMENTED** [review]
Rather than returning an error here why not implement a default behavior for unrecognized group versions. As the v1 API implements immutability we should consider making this the default behavior for all group versions that are not specifically known to be mutable.

**Comment by janetkuo** [comment]
> Rather than returning an error here why not implement a default behavior for unrecognized group versions. As the v1 API implements immutability we should consider making this the default behavior for all group versions that are not specifically known to be mutable.

I looked at the original PR and a default behavior wasn't implemented because we wanted to have each new group version make their own decision explicitly.

**Comment by janetkuo** [comment]
Since we're more confident about the selector immutability strategy now, we can make it a default behavior for all future versions, if not returning an error here. Either way seems fine to me, as long as we don't `panic`.

**Comment by janetkuo** [comment]
Pushed another commit that makes "selector immutability" the default behavior. @liggitt @kow3ns PTAL

**Comment by liggitt** [comment]
I'm ok with this, as long as we're confident that relaxing immutability in the future won't be considered a breaking API change

**Comment by liggitt** [comment]
(needs squash)

**Comment by kow3ns** [comment]
might also need hack/update-bazel.sh

**Comment by janetkuo** [comment]
Squashed and updated bazel. Re-applying tag.

**Pull request merged**